File: exam-conductor/archiveDCR/libs/exampen-proto/python/exampen_proto/score.py
"""Score domain models: score projections, overrides, audit history, workflow."""

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel

from .enums import ScoreEventType, ScoreLifecycleState


class StepScore(BaseModel):
    """Score for a single step within a question."""

    label: str
    awarded: float
    max: float


class QuestionScore(BaseModel):
    """Score projection for one question."""

    question_id: str
    ai_score: float
    current_score: float
    max_score: float
    confidence: float
    override_reason: Optional[str] = None
    step_scores: Optional[list[StepScore]] = None


class StudentScoreDetail(BaseModel):
    """Full score projection for one student in an exam."""

    exam_id: UUID
    student_id: str
    total_score: float
    max_score: Optional[float] = None
    lifecycle_state: ScoreLifecycleState
    published_at: Optional[datetime] = None
    objection_window_closes_at: Optional[datetime] = None
    questions: list[QuestionScore]


class ScoreOverrideRequest(BaseModel):
    """Teacher override for a single question score."""

    teacher_id: str
    new_score: float
    reason: str


class ScoreHistoryItem(BaseModel):
    """Single entry in the score audit event stream."""

    event_id: str
    event_type: ScoreEventType
    old_value: float
    new_value: float
    actor_id: str
    reason: Optional[str] = None
    created_at: datetime


class FinalizeRequest(BaseModel):
    """Request to finalize reviewed scores for an exam."""

    actor_id: str


class PublishRequest(BaseModel):
    """Request to publish finalized scores and open objection window."""

    actor_id: str
    objection_window_days: int


class WorkflowStateResponse(BaseModel):
    """Result of a score workflow state transition."""

    exam_id: UUID
    lifecycle_state: ScoreLifecycleState
    changed_at: datetime


class ScoreUpdatedEvent(BaseModel):
    """NATS event: score projection changed."""

    event_id: str
    event_type: str = "score.updated"
    event_version: str = "1.0.0"
    occurred_at: datetime
    exam_id: UUID
    student_id: str
    question_id: Optional[str] = None
    lifecycle_state: ScoreLifecycleState
    total_score: float
    previous_total_score: Optional[float] = None
    reason: ScoreEventType

File: exam-conductor/archiveDCR/libs/exampen-proto/typescript/src/score.ts
/** Score domain types: projections, overrides, audit history, workflow. */

import type { ScoreEventType, ScoreLifecycleState } from "./enums";

/** Score for a single step within a question. */
export interface StepScore {
  label: string;
  awarded: number;
  max: number;
}

/** Score projection for one question. */
export interface QuestionScore {
  question_id: string;
  ai_score: number;
  current_score: number;
  max_score: number;
  confidence: number;
  override_reason?: string;
  step_scores?: StepScore[];
}

/** Full score projection for one student in an exam. */
export interface StudentScoreDetail {
  exam_id: string;
  student_id: string;
  total_score: number;
  max_score?: number;
  lifecycle_state: ScoreLifecycleState;
  published_at?: string;
  objection_window_closes_at?: string;
  questions: QuestionScore[];
}

/** Teacher override for a single question score. */
export interface ScoreOverrideRequest {
  teacher_id: string;
  new_score: number;
  reason: string;
}

/** Single entry in the score audit event stream. */
export interface ScoreHistoryItem {
  event_id: string;
  event_type: ScoreEventType;
  old_value: number;
  new_value: number;
  actor_id: string;
  reason?: string;
  created_at: string;
}

/** Request to finalize reviewed scores for an exam. */
export interface FinalizeRequest {
  actor_id: string;
}

/** Request to publish finalized scores and open objection window. */
export interface PublishRequest {
  actor_id: string;
  objection_window_days: number;
}

/** Result of a score workflow state transition. */
export interface WorkflowStateResponse {
  exam_id: string;
  lifecycle_state: ScoreLifecycleState;
  changed_at: string;
}

/** NATS event: score projection changed. */
export interface ScoreUpdatedEvent {
  event_id: string;
  event_type: "score.updated";
  event_version: "1.0.0";
  occurred_at: string;
  exam_id: string;
  student_id: string;
  question_id?: string;
  lifecycle_state: ScoreLifecycleState;
  total_score: number;
  previous_total_score?: number;
  reason: ScoreEventType;
}
